### ISSUE TITLE
separate thread for ros::spin causes GUI to crash 

### ISSUE THREAD

**Issue by torydebra** [ADVRHumanoids/rosee_gui#1]
On branch [rosActions](https://github.com/ADVRHumanoids/rosee_gui/tree/rosActions), we have two blocking functions that must work togheter: the exec of QT and the spin of ROS (to get the callbacks of sendGoal). Now I solved creating a thread that run spin: 
~~~C++
void rosSpin_func() {
    ros::spin();
}

int main(int argc, char **argv)
{
    QApplication app (argc, argv);

    ros::init (argc, argv, "rosee_GUI");
    ros::NodeHandle nh;

    Window window(&nh);
    window.show();
    
    //TODO IS IT SAFE?? IS IT CORRECT?
    //BUG with this, the gui crash when we send an action and the previous one
    // was not completed
    std::thread rosSpinner(rosSpin_func);
    int appReturn = app.exec();

    //app.exec is blocking
    rosSpinner.join();
    ros::shutdown();

    return appReturn;
}
~~~

I suppose that it is not so correct. 
At the moment things works but when, I send another command while the previous one is stll running, the GUI crashes.

**Comment by torydebra** [comment]
I found out that this may happend also when receiving a feedback, not only if we send new command. Anyway both crashes happen only sometimes

**Comment by alaurenzi** [comment]
My suggestion is to create a Qt Timer `QTimer`, and use it to call `ros::spinOnce` at a predefined frequency with the Qt signal/slot mechanism.

Your approach is also correct (and in principle it is always better to separate gui and logic into different threads), but you have to take care of data races between Qt and ROS.

**Comment by torydebra** [comment]
Thanks! I will try

**Comment by torydebra** [comment]
Solved as suggested in 38069b3e4d0c59c14f935134786edd032d65409a. Now it seems to not crash anymore. Thanks @alaurenzi  !
